src/testers.rs

```diff
@@ -1,30 +1,123 @@
 use errors::{TeraResult, TeraError};
-use serde_json::value::{Map, Value};
-use parser::Node;
-
-use std::collections::LinkedList;
-
-// TODO: Don't expose the AST to tester functions.
-pub type TesterFn = fn(context: &Map<String, Value>,
-                       value: &Node,
-                       params: LinkedList<Value>)
-                       -> TeraResult<bool>;
-
-/// Returns true if `value` is defined in the given context. Otherwise, returns
-/// false.
-pub fn defined(context: &Map<String, Value>, value: &Node, params: LinkedList<Value>)
-        -> TeraResult<bool> {
-    if params.len() != 0 {
-        return Err(TeraError::TestError("defined".to_string(),
-            "defined should not be called with parameters".to_string()))
+use serde_json::value::{Value};
+use context::ValueNumber;
+
+
+
+pub type TesterFn = fn(
+    value: Option<Value>,
+    params: Vec<Value>
+) -> TeraResult<bool>;
+
+
+// Some helper functions to remove boilerplate with tester error handling
+fn number_args_allowed(name: &str, max: usize, args_len: usize) -> TeraResult<()> {
+    if args_len > max {
+        return Err(TeraError::TestError(
+            name.to_string(),
+            format!("{} should not be called with parameters", name)
+        ))
     }
 
-    let name = match *value {
-        Node::Identifier { ref name, .. } => name,
-        _ => return Err(TeraError::TestError("defined".to_string(),
-                "defined can only be called on identifiers".to_string()))
-    };
+    Ok(())
+}
+
+// Called to return an error when unwrapping the value and realising there's nothing
+fn value_is_undefined(name: &str) -> TeraResult<bool> {
+    Err(TeraError::TestError(
+        name.to_string(),
+        format!("{} was called on an undefined expression", name)
```

Comment:
And the name should be in the error message: "`{}` was called on the variable `{}`, which is undefined."